### ISSUE TITLE
 ActiveRole the onLeaderHeartbeat uses wrong context for transition

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#4164]
In the ActiveRole the onLeaderHeartbeat tries to transition without switching the context.
This is not really an issue since it doesn't break anything it is just not nice to have it in the log.
The Activerole onLeaderHeartbeat implementation is also just in use when the node is Candidate, so it is really unlikely.

```
I 2019-11-12T12:06:38.171859086Z 2019-11-12 12:06:38.171 [] [raft-partition-group-system-8] INFO  io.atomix.protocols.raft.partition.impl.RaftPartitionServer - Starting server for partition PartitionId{id=5, group=raft-atomix}
 
I 2019-11-12T12:06:38.171919552Z 2019-11-12 12:06:38.170 [] [raft-server-4-raft-atomix-partition-8] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-atomix-partition-8} - Transitioning to FOLLOWER
 
I 2019-11-12T12:06:39.449518990Z 2019-11-12 12:06:39.448 [] [raft-server-4-raft-atomix-partition-5] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-atomix-partition-5} - Transitioning to CANDIDATE
 
I 2019-11-12T12:06:39.450871451Z 2019-11-12 12:06:39.450 [] [raft-partition-group-system-8] INFO  io.atomix.protocols.raft.partition.impl.RaftPartitionServer - Starting server for partition PartitionId{id=3, group=raft-atomix}
 
I 2019-11-12T12:06:39.452037111Z 2019-11-12 12:06:39.451 [] [raft-server-4-raft-atomix-partition-5] INFO  io.atomix.protocols.raft.roles.CandidateRole - RaftServer{raft-atomix-partition-5}{role=CANDIDATE} - Starting election
 
I 2019-11-12T12:06:39.469407602Z 2019-11-12 12:06:39.468 [] [raft-server-4-raft-atomix-partition-5] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-atomix-partition-5} - Transitioning to FOLLOWER
 
I 2019-11-12T12:06:39.470071477Z 2019-11-12 12:06:39.469 [] [raft-server-4-raft-atomix-partition-5-heartbeat] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-atomix-partition-5} - Found leader 1
 
I 2019-11-12T12:06:39.479566674Z 2019-11-12 12:06:39.473 [] [raft-server-4-raft-atomix-partition-5-heartbeat] ERROR io.atomix.utils.concurrent.SingleThreadContext - An uncaught exception occurred
 
I 2019-11-12T12:06:39.479593576Z java.lang.IllegalStateException: not on a Catalyst thread
	at com.google.common.base.Preconditions.checkState(Preconditions.java:508) ~[guava-28.1-jre.jar:?]
	at io.atomix.utils.concurrent.ThreadContext.checkThread(ThreadContext.java:71) ~[atomix-utils-3.2.0-alpha8.jar:?]
	at io.atomix.protocols.raft.impl.RaftContext.checkThread(RaftContext.java:750) ~[atomix-raft-3.2.0-alpha8.jar:?]
	at io.atomix.protocols.raft.impl.RaftContext.transition(RaftContext.java:915) ~[atomix-raft-3.2.0-alpha8.jar:?]
	at io.atomix.protocols.raft.roles.ActiveRole.onLeaderHeartbeat(ActiveRole.java:76) ~[atomix-raft-3.2.0-alpha8.jar:?]
	at io.atomix.protocols.raft.impl.RaftContext.lambda$null$35(RaftContext.java:776) ~[atomix-raft-3.2.0-alpha8.jar:?]
	at io.atomix.utils.concurrent.SingleThreadContext$1.lambda$execute$0(SingleThreadContext.java:53) ~[atomix-utils-3.2.0-alpha8.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
 
I 2019-11-12T12:06:41.669784321Z 2019-11-12 12:06:41.668 [] [raft-server-4-raft-atomix-partition-3] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-atomix-partition-3} - Transitioning to FOLLOWER
 
I 2019-11-12T12:06:41.670951037Z 2019-11-12 12:06:41.670 [] [raft-partition-group-system-8] INFO  io.atomix.protocols.raft.partition.impl.RaftPartitionServer - Starting server for partition PartitionId{id=4, group=raft-atomix}
 ```

**Comment by deepthidevaki** [comment]
It is not harmless as we thought before. ` Activerole:onLeaderHeartbeat ` is also used by LeaderRole. It can happen that it first update the term and leader from the new leader info and fail to transition to follower. Then it is still in the LeaderRole but with incorrect new term.

**Comment by deepthidevaki** [comment]
Closed by #4219
